resources/indexer.py

```diff
@@ -314,20 +312,209 @@ def _get_i3_part_file_metadata(file, site, run_meta_xml=None, gaps_dict=None, gc
-    if ".i3" in file.name:
-        try:  # check if last char of filename (w/o extension) is an int
-            int(file.name.split('.i3')[0][-1])
-            return True
-        except ValueError:
-            return False
-    return False
+        return metadata
+
+    def _parse_filepath(self):
+        """Set the year, run, subrun, and part from the file name."""
+        self.run = I3FileMetadata.parse_run_number(self.file)
+
+        # Ex: Level2_IC86.2017_data_Run00130567_Subrun00000000_00000280.i3.zst
+        # Ex: Level2pass2_IC79.2010_data_Run00115975_Subrun00000000_00000055.i3.zst
+        if re.match(r'(.*)(\.20\d{2})_data_Run[0-9]+_Subrun[0-9]+_[0-9]+(.*)', self.file.name):
+            y = self.file.name.split('.')[1]
+            self.season_year = int(y.split('_')[0])
+            s = self.file.name.split('Subrun')[1]
+            self.subrun = int(s.split('_')[0])
+            p = self.file.name.split('_')[-1]
+            self.part = int(p.split('.')[0])
+
+        # Ex: Level2_IC86.2016_data_Run00129004_Subrun00000316.i3.bz2
+        elif re.match(r'(.*)(\.20\d{2})_data_Run[0-9]+_Subrun[0-9]+(.*)', self.file.name):
+            y = self.file.name.split('.')[1]
+            self.season_year = int(y.split('_')[0])
+            self.subrun = 0
+            p = self.file.name.split('Subrun')[1]
+            self.part = int(p.split('.')[0])
+
+        # Ex: Level2_IC86.2011_data_Run00119221_Part00000126.i3.bz2
+        elif re.match(r'(.*)(\.20\d{2})_data_Run[0-9]+_Part[0-9]+(.*)', self.file.name):
+            y = self.file.name.split('.')[1]
+            self.season_year = int(y.split('_')[0])
+            self.subrun = 0
+            p = self.file.name.split('Part')[1]
+            self.part = int(p.split('.')[0])
+
+        # Ex: Level2a_IC59_data_Run00115968_Part00000290.i3.gz
+        elif re.match(r'(.*)(_IC.*)_data_Run[0-9]+_Part[0-9]+(.*)', self.file.name):
+            n = self.file.name.split('IC')[1]
+            strings = n.split('_')[0]
+            self.season_year = IceCubeSeason.name_to_year(f"IC{strings}")
+            self.subrun = 0
+            p = self.file.name.split('Part')[1]
+            self.part = int(p.split('.')[0])
+
+        # Ex: Level2_All_Run00111562_Part00000046.i3.gz
+        elif re.match(r'(.*)_Run[0-9]+(.*)_Part[0-9]+(.*)', self.file.name):
+            self.season_year = None
+            self.subrun = 0
+            p = self.file.name.split('Part')[1]
+            self.part = int(p.split('.')[0])
+
+        else:
+            raise Exception(f"Filename not in a known L2 file format, {self.file.name}.")
+
+    @staticmethod
+    def is_file(file):
+        """True if the file is in the [...#].i3[...] file format."""
+        # Ex: Level2_IC86.2017_data_Run00130484_Subrun00000000_00000188.i3.zst
+        # check if last char of filename (w/o extension) is an int
+        return (".i3" in file.name) and (file.name.split('.i3')[0][-1]).isdigit()
```

Comment:
You've got the `file` attribute from the grandparent superclass, so make this a non-static method perhaps?
Or refactor it out of the class into a stand-alone function?

Reply:
You mention this in https://github.com/WIPACrepo/file_catalog/pull/52#discussion_r376175831